include/boost/geometry/strategies/transform/matrix_transformers.hpp

```diff
@@ -41,6 +44,102 @@ namespace boost { namespace geometry
 namespace strategy { namespace transform
 {
 
+namespace detail { namespace matrix_transformer
+{
+
+template
+<
+    typename CalculationType,
+    typename P2,
+    std::size_t Dimension2,
+    std::size_t Dimension2Count
+>
+struct set_point_from_vec
+{
+    typedef CalculationType ct;
+    static inline void apply(P2& p2, qvm::vec<ct, Dimension2 + 1>& p2temp)
+    {
+        typedef typename geometry::coordinate_type<P2>::type ct2;
+        set<Dimension2Count>(p2, boost::numeric_cast<ct2>(qvm::A<Dimension2Count>(p2temp)));
+        set_point_from_vec<CalculationType, P2, Dimension2, Dimension2Count-1>::apply(p2, p2temp);
+    }
+};
+
+template
+<
+    typename CalculationType,
+    typename P2,
+    std::size_t Dimension2
+>
+struct set_point_from_vec<CalculationType, P2, Dimension2, -1>
+{
+    typedef CalculationType ct;
+    static inline void apply(P2& p2, qvm::vec<ct, Dimension2 + 1>& p2temp) {}
+};
+
+template
+<
+    typename CalculationType,
+    typename P1,
+    std::size_t Dimension1,
+    std::size_t Dimension1Count
+>
+struct set_vec_from_point
+{
+    typedef CalculationType ct;
+    static inline void apply(const P1& p1, qvm::vec<ct, Dimension1 + 1>& p1temp)
+    {
+        qvm::A<Dimension1Count>(p1temp) = get<Dimension1Count>(p1);
+        set_vec_from_point<CalculationType, P1, Dimension1, Dimension1Count - 1>::apply(p1, p1temp);
+    }
+};
+
+template
+<
+    typename CalculationType,
+    typename P1,
+    std::size_t Dimension1
+>
+struct set_vec_from_point<CalculationType, P1, Dimension1, -1>
+{
+    typedef CalculationType ct;
+    static inline void apply(const P1& p1, qvm::vec<ct, Dimension1 + 1>& p1temp) {}
+};
+
+
+template
+<
+    typename CalculationType,
+    std::size_t Dimension1,
+    std::size_t Dimension2
+>
+class matrix_transformer
+{
+protected :
+    typedef CalculationType ct;
+    typedef boost::qvm::mat<ct, Dimension2 + 1, Dimension1 + 1> matrix_type;
+    matrix_type m_matrix;
```

Comment:
AFAIU the matrix stored internally should always he square transformation matrix of the output dimension, so for 2d output the matrix should be 3x3, for 3d output 4x4. Am I right @barendgehrels @vissarion ?

If that's correct then two DimensionX template arguments are not needed, you need only one.

Also see the comment below about transformer from 3d to 2d.

Reply:
What if you want to go from 3d to 2d?

Reply:
If I am not missing something, it seems that you need a 4x3 transformation matrix to map from 3d to 2d, so not always square.

Reply:
AFAIU @barendgehrels was the original author so I'd prefer if he expressed his opinion thoughts regarding it assuming he has some. @tinko92 how badly would you like this to be released with 1.71?

Reply:
I need more time to review this. Changing matrix dimensions does not look good to me. Does it not break anything?
I come back to this next week.

Reply:
@awulkiew Not so badly that the review would be rushed, I can use private patches in my codebase for another release cycle. I also understand that 1.71 has just been closed for major changes.

@barendgehrels This change in matrix dimensions is, I think, the single change to visible behavior in this whole pull request (I hope I was sufficiently transparent about this in my initial pull request). It can break code that takes the matrix from a 3d to 2d transformer and specifically expects it to be 3 by 3. The method that returns the internal matrix is not documented, it is not covered by test cases but it is a public method and it is shown in an example.

Here is my rationale for proposing this breaking change:
The matrix transformers are the base class of rotation, scaling, and translation, they are applied as matrix transformation on homogenous coordinates, in other words: they represent affine transformations. Two things about this:

1. Affine transformations can be composed via matrix multiplication, a useful property demonstrated in example 06b ( https://github.com/boostorg/geometry/blob/f60ce3691f5941855ea8b503629b4d509156bfeb/example/06_b_transformation_example.cpp#L147 ). Currently one could not compose a 3d->3d transformation with a 3d->2d transformation, because it is not sensible to multiply a 3x3 matrix and a 4x4 matrix. With these patches, it will work because it is possible to multiply a 3x4 matrix and a 4x4 matrix to get a 3x4 matrix like it makes sense to compose a 3d->3d transformation and a 3d->2d transformation to get a 3d->2d transformation.

2. A 3x4 matrix is necessary to represent all possible affine transformations from 3d to 2d. The current implementation cannot represent, for example, the transformation: (x, y, z) -> (2x + z + 1, 3y + z + 1). My proposed implementation does support this case.

Reply:
> It can break code that takes the matrix from a 3d to 2d transformer and specifically expects it to be 3 by 3. 

@tinko92 You're talking about the constructor taking `matrix_type` correct? If that's the case I wouldn't care about it that much since it's a new addition done by you which was released with 1.71. Before we only had constructors taking the elements of the matrix and for 3->2 case the constructor was not changed, it still takes 3x3=9 elements. So we are backward compatible.

Would it make sense to add constructor taking 3x4=12 elements and for symmetry another one taking 3x3 matrix? It would be possible to support two generic matrix types having different sizes with `enable_if`.

Reply:
@awulkiew This is true for the constructor, but there is also this method: https://github.com/boostorg/geometry/blob/5a13eea5edbad79c9512dd66ac18e6c0b553b489/include/boost/geometry/strategies/transform/matrix_transformers.hpp#L111 . I think this one was available before 1.71 and it will return another matrix type after the change for 3->2 transformers.

If you consider it sensible, I could definitely implemented the additional constructors that you're proposing.

Reply:
Right, so we have to merge ASAP and list it as a breaking change in release notes.